// angular-slickgrid/esm2015/app/modules/angular-slickgrid/models/fieldType.enum.js
export var FieldType;
(function (FieldType) {
    /** unknown type */
    FieldType[FieldType["unknown"] = 0] = "unknown";
    /** string type */
    FieldType[FieldType["string"] = 1] = "string";
    /** boolean type (true/false) */
    FieldType[FieldType["boolean"] = 2] = "boolean";
    /** integer number type (1,2,99) */
    FieldType[FieldType["integer"] = 3] = "integer";
    /** float number (with decimal) type */
    FieldType[FieldType["float"] = 4] = "float";
    /** number includes Integer and Float */
    FieldType[FieldType["number"] = 5] = "number";
    /** new Date(), javascript Date object */
    FieldType[FieldType["date"] = 6] = "date";
    /** Format: 'YYYY-MM-DD' => 2001-02-28 */
    FieldType[FieldType["dateIso"] = 7] = "dateIso";
    /** Format: 'YYYY-MM-DDTHH:mm:ss.SSSZ' => 2001-02-28T14:00:00.123Z */
    FieldType[FieldType["dateUtc"] = 8] = "dateUtc";
    /** new Date(), javacript Date Object with Time */
    FieldType[FieldType["dateTime"] = 9] = "dateTime";
    /** Format: 'YYYY-MM-DD HH:mm:ss' => 2001-02-28 14:01:01 */
    FieldType[FieldType["dateTimeIso"] = 10] = "dateTimeIso";
    /** Format: 'YYYY-MM-DD h:mm:ss a' => 2001-02-28 11:01:01 pm */
    FieldType[FieldType["dateTimeIsoAmPm"] = 11] = "dateTimeIsoAmPm";
    /** Format: 'YYYY-MM-DD h:mm:ss A' => 2001-02-28 11:01:01 PM */
    FieldType[FieldType["dateTimeIsoAM_PM"] = 12] = "dateTimeIsoAM_PM";
    /** Format: 'YYYY-MM-DD HH:mm' => 2001-02-28 14:01 */
    FieldType[FieldType["dateTimeShortIso"] = 13] = "dateTimeShortIso";
    /** Format (Euro): 'DD/MM/YYYY' => 28/02/2001 */
    FieldType[FieldType["dateEuro"] = 14] = "dateEuro";
    /** Format (Euro): 'D/M/YY' => 28/2/12 */
    FieldType[FieldType["dateEuroShort"] = 15] = "dateEuroShort";
    /** Format (Euro): 'DD/MM/YYYY HH:mm' => 28/02/2001 13:01 */
    FieldType[FieldType["dateTimeShortEuro"] = 16] = "dateTimeShortEuro";
    /** Format (Euro): 'DD/MM/YYYY HH:mm:ss' => 02/28/2001 13:01:01 */
    FieldType[FieldType["dateTimeEuro"] = 17] = "dateTimeEuro";
    /** Format (Euro): 'DD/MM/YYYY hh:mm:ss a' => 28/02/2001 11:01:01 pm */
    FieldType[FieldType["dateTimeEuroAmPm"] = 18] = "dateTimeEuroAmPm";
    /** Format (Euro): 'DD/MM/YYYY hh:mm:ss A' => 28/02/2001 11:01:01 PM */
    FieldType[FieldType["dateTimeEuroAM_PM"] = 19] = "dateTimeEuroAM_PM";
    /** Format (Euro): 'D/M/YY H:m:s' => 28/2/14 14:1:2 */
    FieldType[FieldType["dateTimeEuroShort"] = 20] = "dateTimeEuroShort";
    /** Format (Euro): 'D/M/YY h:m:s a' => 28/2/14 1:2:10 pm */
    FieldType[FieldType["dateTimeEuroShortAmPm"] = 21] = "dateTimeEuroShortAmPm";
    /** Format (Euro): 'D/M/YY h:m:s A' => 28/2/14 14:1:1 PM */
    FieldType[FieldType["dateTimeEuroShortAM_PM"] = 22] = "dateTimeEuroShortAM_PM";
    /** Format: 'MM/DD/YYYY' => 02/28/2001 */
    FieldType[FieldType["dateUs"] = 23] = "dateUs";
    /** Format: 'M/D/YY' => 2/28/12 */
    FieldType[FieldType["dateUsShort"] = 24] = "dateUsShort";
    /** Format: 'MM/DD/YYYY HH:mm' => 02/28/2001 13:01 */
    FieldType[FieldType["dateTimeShortUs"] = 25] = "dateTimeShortUs";
    /** Format: 'MM/DD/YYYY HH:mm:ss' => 02/28/2001 13:01:01 */
    FieldType[FieldType["dateTimeUs"] = 26] = "dateTimeUs";
    /** Format: 'MM/DD/YYYY hh:mm:ss a' => 02/28/2001 11:01:01 pm */
    FieldType[FieldType["dateTimeUsAmPm"] = 27] = "dateTimeUsAmPm";
    /** Format: 'MM/DD/YYYY hh:mm:ss A' => 02/28/2001 11:01:01 PM */
    FieldType[FieldType["dateTimeUsAM_PM"] = 28] = "dateTimeUsAM_PM";
    /** Format: 'M/D/YY H:m:s' => 2/28/14 14:1:2 */
    FieldType[FieldType["dateTimeUsShort"] = 29] = "dateTimeUsShort";
    /** Format: 'M/D/YY h:m:s a' => 2/28/14 1:2:10 pm */
    FieldType[FieldType["dateTimeUsShortAmPm"] = 30] = "dateTimeUsShortAmPm";
    /** Format: 'M/D/YY h:m:s A' => 2/28/14 14:1:1 PM */
    FieldType[FieldType["dateTimeUsShortAM_PM"] = 31] = "dateTimeUsShortAM_PM";
    /** complex object with various properties */
    FieldType[FieldType["object"] = 32] = "object";
})(FieldType || (FieldType = {}));
